include/sparrow/any_data.hpp

```diff
@@ -0,0 +1,349 @@
+    {
+    public:
+
+        explicit any_data(T* data);
+
+        template <mpl::unique_ptr_or_derived U>
+            requires std::is_same_v<typename U::element_type, T>
+        explicit any_data(U data);
+
+        template <mpl::shared_ptr_or_derived U>
+            requires std::is_same_v<typename U::element_type, T>
+        explicit any_data(U data);
+
+
+        explicit any_data(T data)
+            requires(!mpl::smart_ptr_and_derived<T>);
+
+        T* get();
+        const T* get() const;
+
+        // Returns the data as a reference to the type U.
+        // The type U must be the same as the type of the data. Otherwise std::bad_any_cast exception is
+        // thrown.
+        template <class U>
+        U get_data();
+
+        template <class U>
+        const U get_data() const;
+
+        [[nodiscard]] bool owns_data() const noexcept;
+
+        std::type_index type_id() const;
+
+    private:
+
+        std::any m_owner;
+        T* m_raw_ptr = nullptr;
+    };
+
+    template <typename T>
+    any_data<T>::any_data(T* data)
+        : m_raw_ptr(data)
+    {
+    }
+
+    template <typename T>
+    template <mpl::unique_ptr_or_derived U>
+        requires std::is_same_v<typename U::element_type, T>
+    any_data<T>::any_data(U data)
+        : m_owner(std::move(value_ptr<typename U::element_type, typename U::deleter_type>(std::move(data))))
+        , m_raw_ptr(std::any_cast<value_ptr<typename U::element_type, typename U::deleter_type>&>(m_owner).get())
+    {
+    }
+
+    template <typename T>
+    template <mpl::shared_ptr_or_derived U>
+        requires std::is_same_v<typename U::element_type, T>
+    any_data<T>::any_data(U data)
+        : m_owner(std::move(data))
+        , m_raw_ptr(std::any_cast<U&>(m_owner).get())
+    {
+    }
+
+    template <typename T>
+    any_data<T>::any_data(T data)
+        requires(!mpl::smart_ptr_and_derived<T>)
+        : m_owner(std::move(data))
+        , m_raw_ptr(&std::any_cast<T&>(m_owner))
+    {
+    }
+
+    template <typename T>
+    T* any_data<T>::get()
+    {
+        return m_raw_ptr;
+    }
+
+    template <typename T>
+    const T* any_data<T>::get() const
+    {
+        return m_raw_ptr;
+    }
+
+    template <typename T>
+    template <class U>
+    U any_data<T>::get_data()
+    {
+        return std::any_cast<U>(m_owner);
+    }
+
+    template <typename T>
+    template <class U>
+    const U any_data<T>::get_data() const
+    {
+        return std::any_cast<U>(m_owner);
+    }
+
+    template <typename T>
+    bool any_data<T>::owns_data() const noexcept
+    {
+        return m_owner.has_value();
+    }
+
+    template <typename T>
+    std::type_index any_data<T>::type_id() const
+    {
+        return m_owner.type();
+    }
+
+    /**
+     * @brief A class that can own or not a container of objects and expose them as raw pointers.
+     *
+     * @tparam T The type of the pointers.
+     */
+    template <typename T>
+    class any_data_container
```

Comment:
For readability, it would be nive to have the declaratio of this class following that of `any_data`, and then the implementations.

Reply:
Done